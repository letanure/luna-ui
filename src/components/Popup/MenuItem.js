import PropTypes from 'prop-types';
import React from 'react';
import styled from 'styled-components';

/* eslint-disable no-empty-function */
/**
 * MenuItem Component.
 * ```html
 *  <Popup.NavContextMenu id="2" separator kind="bg" text="EN">
 *    <Popup.MenuItem id="mi2.1">EN</Popup.MenuItem>
 *    <Popup.MenuItem id="mi2.2">DE</Popup.MenuItem>
 *  </Popup.NavContextMenu>
 * ```
 */
class MenuItemComponent extends React.PureComponent {
  render() {
    const { className } = this.props;

    const MenuItem = styled.li`
      transition: all 0.25s linear;
      padding: 8px 30px 8px 14px;
      font-size: 14px;
      line-height: 1.79;
      text-align: left;
      color: #333333;
      cursor: pointer;
      &:hover {
        background-color: #f0f0f0;
      }
    `;
    return (
      <MenuItem className={className} onClick={this.props.onClick}>{this.props.children}</MenuItem>
    );
  }
}

MenuItemComponent.propTypes = {
  onClick: PropTypes.func.isRequired,
  children: PropTypes.node,
  className: PropTypes.string
};

MenuItemComponent.defaultProps = {
  children: null,
  className: '',
  onClick: () => {}
};

export default MenuItemComponent;
