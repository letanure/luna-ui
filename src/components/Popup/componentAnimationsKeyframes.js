import { keyframes } from 'styled-components';

export const componentAnimationsKeyframes = {
  contextMenu: keyframes`
    from {
      opacity: 0;
      visibility: hidden;
      transform: translate(-115px, -30px);
    }

    to {
      opacity: 1;
      visibility: visible;
      transform: translate(-125px, -30px);
    }
  `,
  navMenu: keyframes`
    from {
      opacity: 0;
      visibility: hidden;
      transform: translate(0, -30px);
    }

    to {
      opacity: 1;
      visibility: visible;
      transform: translate(0, 1px);
    }
  `
};
