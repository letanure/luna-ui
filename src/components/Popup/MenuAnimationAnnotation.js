import { componentAnimationsKeyframes } from './componentAnimationsKeyframes';

export function MenuAnimation(target) {
  target.prototype._getAnimation = function(animationTarget, inverse = false) {
    if (this.state.active && !inverse) {
      return `animation: ${componentAnimationsKeyframes[animationTarget]} .25s ease-in;`;
    }
    return `animation: ${componentAnimationsKeyframes[animationTarget]} .25s reverse ease-in; `;
  };

  target.prototype._getElementAnimationClass = function(animatedElement, firstLoad) {
    const animatedElemetClasses = (animatedElement !== undefined) && animatedElement.classList.toString();

    if (this.state.active) {
      return 'elementIn';
    } else if ((animatedElemetClasses && animatedElemetClasses.indexOf('elementIn') !== -1) && !firstLoad) {
      return 'elementOut';
    }
    return '';
  };
}
