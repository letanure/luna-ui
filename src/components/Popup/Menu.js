import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { MenuAnimation } from './MenuAnimationAnnotation';

let fistLoad = true;

@MenuAnimation
class MenuComponent extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = { active: props.active };
  }

  componentDidMount() {
    fistLoad = false;
    // eslint-disable-next-line no-undef
    document.addEventListener('click', this.props.onClickOutside);
  }
  componentWillUnmount() {
    // eslint-disable-next-line no-undef
    document.removeEventListener('click', this.props.onClickOutside);
  }

  render() {
    const Menu = styled.ul`
      width: 100%;
      min-width: 120px;
      z-index: 999;
      list-style: none;
      position: absolute;
      border-radius: 2px;
      background-color: #ffffff;
      border: solid 1px #d5d5d5;

      // animation
      opacity: 0;
      visibility: hidden;
      transform: translate(-115px, -30px);

      &.elementIn {
        ${this._getAnimation('contextMenu')}
        animation-fill-mode: forwards;
      }

      &.elementOut {
        ${this._getAnimation('contextMenu', true)}
        animation-fill-mode: forwards;
      }
    `;

    /* eslint-disable no-return-assign */
    return <Menu
      innerRef={(comp) => this.animatedElement = comp}
      className={this._getElementAnimationClass(this.animatedElement, fistLoad)}
    >{this.props.children}
    </Menu>;
  }
}

MenuComponent.propTypes = {
  active: PropTypes.bool,
  children: PropTypes.node,
  onClickOutside: PropTypes.func.isRequired
};

MenuComponent.defaultProps = {
  active: false
};

export default MenuComponent;
