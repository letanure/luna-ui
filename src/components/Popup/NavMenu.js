import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { MenuAnimation } from './MenuAnimationAnnotation';

let fistLoad = true;

/**
 * NavMenu (Subcomponent) Component.
 * ```html
 *  <NavMenu active={this.state.isPopupMenuActive}
 *   onClickOutside={this.onClickOutside.bind(this)}>{this.props.children}</NavMenu>
 * ```
 */
@MenuAnimation
class NavMenuComponent extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = { active: props.active };
    this.isLastItem = props.isLastItem;
  }

  // eslint-disable-next-line class-methods-use-this
  onClick(event) {
    if (event) {
      event.stopPropagation();
    }
  }

  componentDidMount() {
    fistLoad = false;

    // eslint-disable-next-line no-undef
    document.addEventListener('click', this.props.onClickOutside);
  }

  componentWillUnmount() {
    // eslint-disable-next-line no-undef
    document.removeEventListener('click', this.props.onClickOutside);
  }

  render() {
    const { className } = this.props;

    const NavMenu = styled.ul`
      width: 100%;
      z-index: 999;
      min-width: 120px;
      list-style: none;
      position: absolute;
      border-radius: 2px;
      border: solid 1px #d5d5d5;
      background-color: #ffffff;

      ${this.isLastItem ? 'right: 0;' : ''}

      // animation
      opacity: 0;
      visibility: hidden;
      transform: translate(0, -30px);

      &.elementIn {
        ${this._getAnimation('navMenu')}
        animation-fill-mode: forwards;
      }

      &.elementOut {
        ${this._getAnimation('navMenu', true)}
        animation-fill-mode: forwards;
      }

    `;

    // eslint-disable-next-line no-return-assign
    return <NavMenu
      innerRef={(comp) => this.animatedElement = comp}
      className={`${this._getElementAnimationClass(this.animatedElement, fistLoad)} ${className}`}
      onClick={this.onClick.bind(this)}
    >{this.props.children}</NavMenu>;
  }
}

NavMenuComponent.propTypes = {
  active: PropTypes.bool,
  isLastItem: PropTypes.bool,
  children: PropTypes.node,
  className: PropTypes.string,
  onClickOutside: PropTypes.func.isRequired
};

NavMenuComponent.defaultProps = {
  active: false,
  isLastItem: false,
  children: null,
  className: ''
};

export default NavMenuComponent;
