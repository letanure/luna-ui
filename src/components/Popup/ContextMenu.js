import React from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';
import FontAwesomeIcon from '@fortawesome/react-fontawesome';

import Menu from './Menu';

class ContextMenuComponent extends React.Component {
  constructor(props) {
    super(props);

    this.firstLoad = true;
    this.state = { active: props.active };
  }

  onClickOutside() {
    if (!this.state.active) {
      return;
    }
    this.toggleMenu();
  }

  toggleMenu() {
    this.setState((prevState) => ({
      active: !prevState.active
    }));

    this.firstLoad = !this.firstLoad;
  }

  _getAnimation(inverse, props) {
    const inactiveBackground = `${!inverse ?
      props.theme.popup.contextMenu.inactiveBg :
      props.theme.popup.contextMenu.activeBg}
    `;

    const activeBackground = `${!inverse ?
      props.theme.popup.contextMenu.activeBg :
      props.theme.popup.contextMenu.inactiveBg}
    `;

    const backgroundAnimationKeyframes = keyframes`
      from {
        color: ${props.theme.popup.contextMenu.inactiveIcon};
        background-color: ${inactiveBackground};
      }

      to {
        color: ${props.theme.popup.contextMenu.activeIcon};
        background-color: ${activeBackground};
      }
    `;

    if (!this.state.active && this.firstLoad) {
      return `background-color: ${inactiveBackground}; color: ${props.theme.popup.contextMenu.inactiveIcon};`;
    } else if (this.state.active && !this.firstLoad) {
      return `animation: ${backgroundAnimationKeyframes} .25s ease-in;`;
    }
    return `animation: ${backgroundAnimationKeyframes} .25s reverse ease-in; `;
  }

  render() {
    const { inverse, className } = this.props;

    const CtxFontAwesomeIcon = styled(FontAwesomeIcon)`
      display: block;
      margin: 2px 0px;
      font-size: 5px;
    `;

    const ContextMenu = styled.div`
      cursor: pointer;
      border-radius: 2px;
      position: relative;
      display: inline-block;
    `;

    const ContextMenuButton = styled.button`
      padding: 2px 6px;
      border: none;
      outline: none;
      cursor: pointer;
      display: flex;
      flex-direction: column;
      font-weight: bold;
      border-radius: 2px;

      ${(props) => this._getAnimation(inverse, props)}
      animation-fill-mode: forwards;
    `;

    return (
      <ContextMenu className={className} onClick={this.toggleMenu.bind(this)}>
        <ContextMenuButton>
          <CtxFontAwesomeIcon icon="circle" />
          <CtxFontAwesomeIcon icon="circle" />
          <CtxFontAwesomeIcon icon="circle" />
        </ContextMenuButton>
        <Menu active={this.state.active} onClickOutside={this.onClickOutside.bind(this)}>{this.props.children}</Menu>
      </ContextMenu>
    );
  }
}

ContextMenuComponent.propTypes = {
  active: PropTypes.bool,
  children: PropTypes.node,
  inverse: PropTypes.bool,
  className: PropTypes.string
};

ContextMenuComponent.defaultProps = {
  active: false,
  inverse: false,
  className: ''
};

export default ContextMenuComponent;
