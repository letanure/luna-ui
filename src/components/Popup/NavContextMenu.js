import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import FontAwesomeIcon from '@fortawesome/react-fontawesome';

import NavMenu from './NavMenu';

/* eslint-disable no-extra-parens, indent, space-infix-ops */

/**
 * NavContextMenu Component.
 * ```html
 *  <Popup.NavContextMenu id="2" separator kind="bg" text="EN">
 *    ...
 *  </Popup.NavContextMenu>
 * ```
 */
class NavContextMenuComponent extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = { isPopupMenuActive: props.active };
  }

  onClickOutside() {
    this.setState({ isPopupMenuActive: false });
  }

  toggleMenu() {
    this.setState((prevState) => ({
      isPopupMenuActive: !prevState.isPopupMenuActive
    }));
  }

  render() {
    const { separator, icon, text, className } = this.props;
    const CtxFontAwesomeIconLeft = styled(FontAwesomeIcon)`
      margin-right: 10px;
      vertical-align: -0.19em;
    `;

    const CtxFontAwesomeIconRight = styled(FontAwesomeIcon)`
      ${text ? 'margin-left: 10px;' : null}
      ${(props) => `color: ${props.theme.popup.navMenu.toggleIcon}`}
    `;

    const NavContextMenu = styled.div`
      cursor: pointer;
      position: relative;
      display: inline-block;
      transition: all .25s linear;

      ${(props) => (separator ?
          `border-right: 1px solid ${props.theme.popup.navMenu.separator};`:
          'border-right: 1px solid transparent;')
}

      ${(props) => (this.state.isPopupMenuActive ?
        `background-color: ${props.theme.popup.navMenu.activeBg};`:
        'background-color: transparent')
}
    `;

    const NavContextMenuButton = styled.button`
      width: 100%;
      height: 40px;
      border: none;
      outline: none;
      z-index: 99999;
      padding: 0 10px;
      cursor: pointer;
      font-size: 12px;
      position: relative;

      ${(props) => (this.state.isPopupMenuActive ?
        `background-color: ${props.theme.popup.navMenu.activeBg};`:
        'background-color: transparent;')
}
    `;

    const customIconComponent = icon && <CtxFontAwesomeIconLeft icon={icon} fixedWidth />;

    const toggleIconComponent = this.state.isPopupMenuActive ?
      <CtxFontAwesomeIconRight icon="caret-up" fixedWidth /> :
      <CtxFontAwesomeIconRight icon="caret-down" fixedWidth />;

    return (
      <NavContextMenu className={className} {...this.props.config} onClick={this.toggleMenu.bind(this)}>
        <NavContextMenuButton>
          {customIconComponent} {text} {toggleIconComponent}
        </NavContextMenuButton>
        <NavMenu
          isLastItem={this.props.lastItem}
          active={this.state.isPopupMenuActive}
          onClickOutside={this.onClickOutside.bind(this)}
        >{this.props.children}</NavMenu>
      </NavContextMenu>
    );
  }
}

NavContextMenuComponent.propTypes = {
  active: PropTypes.bool,
  lastItem: PropTypes.bool,
  text: PropTypes.node,
  separator: PropTypes.bool,
  icon: PropTypes.string,
  className: PropTypes.string,
  children: PropTypes.node,
  config: PropTypes.object
};

NavContextMenuComponent.defaultProps = {
  active: false,
  lastItem: false,
  separator: false,
  icon: undefined,
  text: undefined,
  children: null,
  config: null,
  className: ''
};

export default NavContextMenuComponent;
