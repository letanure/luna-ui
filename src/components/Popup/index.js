import Menu from './Menu';
import NavMenu from './NavMenu';
import MenuItem from './MenuItem';
import ContextMenu from './ContextMenu';
import NavContextMenu from './NavContextMenu';

export default {
  ContextMenu,
  NavContextMenu,
  NavMenu,
  Menu,
  MenuItem
};
