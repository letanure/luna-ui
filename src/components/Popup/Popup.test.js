import React from 'react';
import sinon from 'sinon';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import theme from '../../config/defaultTheme';
import ThemeProvider from '../Theme/index';

import Popup from './index';

describe('<Popup />', () => {
  const onHandleClick = sinon.spy();
  const menuItem = <Popup.MenuItem onClick={onHandleClick}>Menu Item</Popup.MenuItem>;

  const navContextMenu =
    <ThemeProvider theme={theme}>
      <Popup.NavContextMenu>
        {menuItem}
      </Popup.NavContextMenu>
    </ThemeProvider>;

  const navContextMenuSeparatorText =
    <ThemeProvider theme={theme}>
      <Popup.NavContextMenu separator text="My Menu">
        {menuItem}
      </Popup.NavContextMenu>
    </ThemeProvider>;

  const contextMenu =
    <ThemeProvider theme={theme}>
      <Popup.ContextMenu>
        {menuItem}
      </Popup.ContextMenu>
    </ThemeProvider>;

  test('HandleClick', () => {
    const wrapper = mount(menuItem);

    wrapper.find('li').simulate('click');
    expect(onHandleClick).toHaveProperty('callCount', 1);
  });

  test('toMatchSnapshot: navContextMenu', () => {
    const tree = renderer
      .create(navContextMenu)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('toMatchSnapshot: navContextMenuSeparatorText', () => {
    const tree = renderer
      .create(navContextMenuSeparatorText)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('toMatchSnapshot: contextMenu', () => {
    const tree = renderer
      .create(contextMenu)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
