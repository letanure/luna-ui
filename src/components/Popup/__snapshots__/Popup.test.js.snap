// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popup /> toMatchSnapshot: contextMenu 1`] = `
<div>
  <div
    className="sc-jTzLTM bYQBNw"
    onClick={[Function]}
  >
    <button
      className="sc-fjdhpX ijQRov"
    />
    <ul
      className="sc-jzJRlG jszoBt"
    >
      <li
        className="sc-cSHVUG jeBFMY"
        onClick={[Function]}
      >
        Menu Item
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Popup /> toMatchSnapshot: navContextMenu 1`] = `
<div>
  <div
    className="sc-bxivhb cJZKXe"
    onClick={[Function]}
  >
    <button
      className="sc-ifAKCX ffgzDY"
    >
       
       
    </button>
    <ul
      className="  sc-EHOje aVOIR"
      onClick={[Function]}
    >
      <li
        className="sc-bZQynM blVFwR"
        onClick={[Function]}
      >
        Menu Item
      </li>
    </ul>
  </div>
</div>
`;

exports[`<Popup /> toMatchSnapshot: navContextMenuSeparatorText 1`] = `
<div>
  <div
    className="sc-dnqmqq fxUGAf"
    onClick={[Function]}
  >
    <button
      className="sc-iwsKbI rKKYm"
    >
       
      My Menu
       
    </button>
    <ul
      className="  sc-gZMcBi jrNzwC"
      onClick={[Function]}
    >
      <li
        className="sc-gqjmRU iTXUEk"
        onClick={[Function]}
      >
        Menu Item
      </li>
    </ul>
  </div>
</div>
`;
